ERCS/erc-7589.md

```diff
@@ -0,0 +1,480 @@
+    /// @notice Returns the sum of all tokenAmounts granted to the grantee for the given role.
+    /// @param _role The role identifier.
+    /// @param _tokenAddress The token address.
+    /// @param _tokenId The token identifier.
+    /// @param _grantee The user for which the balance is returned.
+    /// @return balance_ The balance of the grantee for the given role.
+    function roleBalanceOf(
+        bytes32 _role,
+        address _tokenAddress,
+        uint256 _tokenId,
+        address _grantee
+    ) external returns (uint256 balance_);
+}
+```
+
+### Metadata Extension
+
+The Roles Metadata extension extends the traditional JSON-based metadata schema of SFTs. Therefore, DApps supporting
+this feature MUST also implement the metadata extension of [ERC-1155](./eip-1155.md). This extension is **optional** and
+allows developers to provide additional information on roles.
+
+Updated Metadata Schema:
+```json
+{
+
+    /** Existing ERC-1155 Metadata **/
+
+    "title": "Token Metadata",
+    "type": "object",
+    "properties": {
+        "name": {
+          "type": "string",
+          "description": "Identifies the asset to which this token represents"
+        },
+        "decimals": {
+          "type": "integer",
+          "description": "The number of decimal places that the token amount should display - e.g. 18, means to divide the token amount by 1000000000000000000 to get its user representation."
+        },
+        "description": {
+          "type": "string",
+          "description": "Describes the asset to which this token represents"
+        },
+        "image": {
+          "type": "string",
+          "description": "A URI pointing to a resource with mime type image/* representing the asset to which this token represents. Consider making any images at a width between 320 and 1080 pixels and aspect ratio between 1.91:1 and 4:5 inclusive."
+        },
+        "properties": {
+          "type": "object",
+          "description": "Arbitrary properties. Values may be strings, numbers, object or arrays."
+        }
+    },
+
+    /** Additional fields for ERC-7589 **/
+
+    "roles": [{
+        "id": {
+            "type": "bytes32",
+            "description": "Identifies the role"
+        },
+        "name": {
+            "type": "string",
+            "description": "Human-readable name of the role"
+        },
+        "description": {
+            "type": "string",
+            "description": "Describes the role"
+        },
+        "inputs": [{
+          "name": {
+            "type": "string",
+            "description": "Human-readable name of the argument"
+          },
+          "type": {
+            "type": "string",
+            "description": "Solidity type, e.g., uint256 or address"
+          }
+        }]
+    }]
+}
+```
+
+The following code snipped is an example of the additional fields described above:
+
+```json
+{
+  // ... Existing SFT Metadata
```

Comment:
I recommend including a full example that validates with both your schema and the original 1155 schema. It'll make it easier for developers to see exactly how the two interact.